File: src/fstar/FStar.Prettyprint.fs
(*
   Copyright 2008-2018 Microsoft Research

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*)

#light "off"
module FStar.Prettyprint
open FStar.ST
open FStar.All

open FStar.Util
open FStar.Parser.ToDocument

module D = FStar.Parser.Driver
module P = FStar.Pprint

type printing_mode =
  | ToTempFile
  | FromTempToStdout
  | FromTempToFile

let temp_file_name f = format1 "%s.print_.fst" f

let generate (m: printing_mode) filenames =
    let parse_and_prettyprint (m: printing_mode) filename =
        let modul, comments = D.parse_file filename in
        let outf =
          match m with
          | FromTempToStdout -> None
          | FromTempToFile ->
            let outf = open_file_for_writing filename in
            Some outf
          | ToTempFile ->
            let outf = open_file_for_writing (temp_file_name filename) in
            Some outf
        in
        let leftover_comments =
            let comments = List.rev comments in
            let doc, comments = modul_with_comments_to_document modul comments in
                            (* TODO : some problem with the F# generated floats *)
            (match outf with
             | Some f -> append_to_file f <| P.pretty_string (float_of_string "1.0") 100 doc
             | None -> P.pretty_out_channel (float_of_string "1.0") 100 doc stdout);
           comments
        in
        let left_over_doc =
          if not (FStar.List.isEmpty leftover_comments) then
            P.concat  [P.hardline ; P.hardline ; comments_to_document leftover_comments]
          else if m = FromTempToStdout then
            // This isn't needed for FromTempToFile, when using `append_to_file` a newline is added to EoF
            P.concat [P.hardline; P.hardline]
          else
            P.empty
        in
        match outf with
        | None ->
          P.pretty_out_channel (float_of_string "1.0") 100 left_over_doc stdout

        | Some outf ->
          append_to_file outf <| P.pretty_string (float_of_string "1.0") 100 left_over_doc;
          close_file outf
    in
    List.iter (parse_and_prettyprint m) filenames

File: src/ocaml-output/FStar_Prettyprint.ml
open Prims
type printing_mode =
  | ToTempFile 
  | FromTempToStdout 
  | FromTempToFile 
let (uu___is_ToTempFile : printing_mode -> Prims.bool) =
  fun projectee -> match projectee with | ToTempFile -> true | uu___ -> false
let (uu___is_FromTempToStdout : printing_mode -> Prims.bool) =
  fun projectee ->
    match projectee with | FromTempToStdout -> true | uu___ -> false
let (uu___is_FromTempToFile : printing_mode -> Prims.bool) =
  fun projectee ->
    match projectee with | FromTempToFile -> true | uu___ -> false
let (temp_file_name : Prims.string -> Prims.string) =
  fun f -> FStar_Compiler_Util.format1 "%s.print_.fst" f
let (generate : printing_mode -> Prims.string Prims.list -> unit) =
  fun m ->
    fun filenames ->
      let parse_and_prettyprint m1 filename =
        let uu___ = FStar_Parser_Driver.parse_file filename in
        match uu___ with
        | (modul, comments) ->
            let outf =
              match m1 with
              | FromTempToStdout -> FStar_Pervasives_Native.None
              | FromTempToFile ->
                  let outf1 =
                    FStar_Compiler_Util.open_file_for_writing filename in
                  FStar_Pervasives_Native.Some outf1
              | ToTempFile ->
                  let outf1 =
                    let uu___1 = temp_file_name filename in
                    FStar_Compiler_Util.open_file_for_writing uu___1 in
                  FStar_Pervasives_Native.Some outf1 in
            let leftover_comments =
              let comments1 = FStar_Compiler_List.rev comments in
              let uu___1 =
                FStar_Parser_ToDocument.modul_with_comments_to_document modul
                  comments1 in
              match uu___1 with
              | (doc, comments2) ->
                  ((match outf with
                    | FStar_Pervasives_Native.Some f ->
                        let uu___3 =
                          FStar_Pprint.pretty_string
                            (FStar_Compiler_Util.float_of_string "1.0")
                            (Prims.of_int (100)) doc in
                        FStar_Compiler_Effect.op_Less_Bar
                          (FStar_Compiler_Util.append_to_file f) uu___3
                    | FStar_Pervasives_Native.None ->
                        FStar_Pprint.pretty_out_channel
                          (FStar_Compiler_Util.float_of_string "1.0")
                          (Prims.of_int (100)) doc FStar_Compiler_Util.stdout);
                   comments2) in
            let left_over_doc =
              if
                Prims.op_Negation
                  (FStar_Compiler_List.isEmpty leftover_comments)
              then
                let uu___1 =
                  let uu___2 =
                    let uu___3 =
                      let uu___4 =
                        FStar_Parser_ToDocument.comments_to_document
                          leftover_comments in
                      [uu___4] in
                    FStar_Pprint.hardline :: uu___3 in
                  FStar_Pprint.hardline :: uu___2 in
                FStar_Pprint.concat uu___1
              else
                if m1 = FromTempToStdout
                then
                  FStar_Pprint.concat
                    [FStar_Pprint.hardline; FStar_Pprint.hardline]
                else FStar_Pprint.empty in
            (match outf with
             | FStar_Pervasives_Native.None ->
                 FStar_Pprint.pretty_out_channel
                   (FStar_Compiler_Util.float_of_string "1.0")
                   (Prims.of_int (100)) left_over_doc
                   FStar_Compiler_Util.stdout
             | FStar_Pervasives_Native.Some outf1 ->
                 ((let uu___2 =
                     FStar_Pprint.pretty_string
                       (FStar_Compiler_Util.float_of_string "1.0")
                       (Prims.of_int (100)) left_over_doc in
                   FStar_Compiler_Effect.op_Less_Bar
                     (FStar_Compiler_Util.append_to_file outf1) uu___2);
                  FStar_Compiler_Util.close_file outf1)) in
      FStar_Compiler_List.iter (parse_and_prettyprint m) filenames